lib/SlackBot.js

```diff
@@ -225,113 +225,253 @@ function Slackbot(configuration) {
+            // include the response channel so that they can be used in
+            // responding to slash commands and outgoing webhooks
             bot.res = res;
 
-            if (payload.type === 'event_callback') {
-                // Receive messages and trigger events from the Events API
-                res.send('');
-                return handleEventsAPI(payload, bot);
-            } else if (payload.callback_id) {
-                // is this an interactive message callback?
-                res.send('');
-                return handleInteractiveMessage(payload, bot);
-            } else if (payload.command) {
-                // this is a slash command
-                return handleSlashCommand(payload, bot);
-            } else if (payload.trigger_word) {
-                return handleOutgoingWebhook(payload, bot);
-            }
+            // pass the payload into Botkit's message handling pipeline!
+            slack_botkit.ingest(bot, payload, res);
+
+
+
+            //
+            // if (payload.type === 'event_callback') {
+            //     // Receive messages and trigger events from the Events API
+            //     res.send('');
+            //     return handleEventsAPI(payload, bot);
+            // } else if (payload.callback_id) {
+            //     // is this an interactive message callback?
+            //     res.send('');
+            //     return handleInteractiveMessage(payload, bot);
+            // } else if (payload.command) {
+            //     // this is a slash command
+            //     return handleSlashCommand(payload, bot);
+            // } else if (payload.trigger_word) {
+            //     return handleOutgoingWebhook(payload, bot);
+            // }
         });
     };
 
+
+
+    // Send a 200 response back to Slack to acknowledge the message.
+    slack_botkit.middleware.ingest.use(function sendResponse(bot, message, res, next) {
+
+        if (res && res.statusCode) {
+            // this is an http response
+            // always send a 200
+            res.status(200);
+
+            // conditionally send a response back to Slack to acknowledge the message.
+            // we do NOT want to respond to incoming webhooks or slash commands
+            // as the response can be used by developers to actually deliver a reply
+            if (!message.command && !message.trigger_word) {
+                res.send('');
+            }
+        }
+        next();
+
+    });
+
+    /* do delivery confirmations for RTM messages */
+    slack_botkit.middleware.ingest.use(function requireDelivery(bot, message, res, next) {
+        if (message.ok != undefined) {
+            // this is a confirmation of something we sent.
+            if (slack_botkit.config.require_delivery) {
+                // loop through all active conversations this bot is having
+                // and mark messages in conversations as delivered = true
+                for (var t = 0; t < slack_botkit.tasks.length; t++) {
```

Comment:
This would be easier to read and less error prone with some temporary variables. Something like:

```
for (var t = 0; t < slack_botkit.tasks.length; t++) {
 var task = slack_botkit.tasks[t];
 if (task.isActive()) {
     for (var c = 0; c < task.convos.length; c++) {
         var convo = task.convos[c];
         for (var s = 0; s < convo.sent.length; s++) {
             var sent = convo.sent[s];
             if (sent.api_response && sent.api_response.id == message.reply_to) {
                 sent.delivered = true;
                 sent.api_response.ts = message.ts;
             }
         }
     }
 }
}
```